API/Repositories/Implementations/ApiMatchRatingStatsRepository.cs

```diff
@@ -3,85 +3,68 @@
-        //             VolatilityAfter = ra.VolatilityAfter,
-        //             IsAdjustment = true,
-        //             Timestamp = ra.Timestamp
-        //         }
-        //     })
-        //     .ToListAsync();
-        //
-        // // Combine data points, converting Match.StartTime and RatingAdjustment.Timestamp to Date for grouping
-        // var combinedDataPoints = matchRatingStats
-        //     .Select(mrs => new { mrs.StartTime.Date, mrs.DataPoint })
-        //     .Concat(ratingAdjustments.Select(ra => new { ra.Timestamp.Date, ra.DataPoint }))
-        //     .GroupBy(x => x.Date)
-        //     .OrderBy(g => g.Key)
-        //     .Select(g => g.Select(x => x.DataPoint).ToList())
-        //     .ToList();
+        // Default date range to min and max values if not provided
+        dateMin ??= DateTime.MinValue;
+        dateMax ??= DateTime.MaxValue;
+
+        // Fetch Match Rating Stats and group by Match.StartTime.Date
+        var adjustments = await _context
+            .RatingAdjustments
+            .Where(ra =>
+                ra.PlayerId == playerId
+                && ra.AdjustmentType != RatingAdjustmentType.Initial
+                && ra.Match!.Tournament.Ruleset == ruleset
+                && ra.Timestamp >= dateMin
+                && ra.Timestamp <= dateMax
+            )
+            .OrderBy(ra => ra.Timestamp)
+            .Select(ra => new
+            {
+                ra.Timestamp,
+                DataPoint = new PlayerRatingChartDataPointDTO
+                {
+                    Name = ra.Match!.Name ?? "<Unknown>",
+                    MatchId = ra.MatchId,
+                    MatchOsuId = ra.Match.OsuId,
+                    MatchCost = _context.PlayerMatchStats
+                        .Where(pms => pms.PlayerId == ra.PlayerId &&
+                                      pms.MatchId == ra.MatchId)
+                        .Select(pms => pms.MatchCost)
+                        .FirstOrDefault(),
+                    RatingBefore = ra.RatingBefore,
+                    RatingAfter = ra.RatingAfter,
+                    VolatilityBefore = ra.VolatilityBefore,
+                    VolatilityAfter = ra.VolatilityAfter,
+                    IsAdjustment = true,
+                    Timestamp = ra.Match.StartTime
+                }
+            }).ToListAsync();
+
+        // Combine data points, converting Match.StartTime and RatingAdjustment.Timestamp to Date for grouping
+        var combinedDataPoints = adjustments
+            .Select(mrs => new { mrs.Timestamp.Date, mrs.DataPoint })
+            .GroupBy(x => x.Date)
+            .OrderBy(g => g.Key)
+            .Select(g => g.Select(x => x.DataPoint).ToList())
+            .ToList();
 
         // Prepare and return the DTO
-        return Task.FromResult(new PlayerRatingChartDTO { ChartData = new List<List<PlayerRatingChartDataPointDTO>>() });
+        return new PlayerRatingChartDTO { ChartData = combinedDataPoints };
```

Comment:
Feel like `PlayerRatingChartDTO` is a redundant type since all it is is a 'container' for the datapoints. Could probably be replaced wherever it's used to:
```cs
List<List<PlayerRatingChartDataPointDTO>> RatingChart { get; init; } = [ [ ] ];
```
or something similar

Reply:
I avoided that originally because to me it feels harder to maintain, and there's a non-zero chance we expand this DTO in the future.